File: code/jl/discount_spec_rad.jl
using LinearAlgebra                         
using PyPlot
using LaTeXStrings
using QuantEcon
PyPlot.matplotlib[:rc]("text", usetex=true) # allow tex rendering
fontsize=18

ρ(A) = maximum(abs(λ) for λ in eigvals(A))  # Spectral radius

function plot_contours(; savefig=false, 
                         figname="../figures/discount_spec_rad.pdf")

    fig, ax = plt.subplots(figsize=(8,4.2))
    grid_size = 50  
    n = 6
    a_vals = LinRange(0.1, 0.95, grid_size)
    s_vals = LinRange(0.1, 0.32, grid_size)
    μ = 0.96

    R = zeros(grid_size, grid_size)
    L = zeros(n, n)

    for (i_a, a) in enumerate(a_vals)
        for (i_s, s) in enumerate(s_vals)
            mc = tauchen(n, a, sqrt(1 - a^2) * s, (1 - a) * μ)
            z_vals, Q = mc.state_values, mc.p
            β_vals = z_vals # max.(z_vals, 0)
            println(minimum(z_vals))
            for i in 1:n
                for j in 1:n
                    L[i, j] = β_vals[i] * Q[i, j]
                end
            end
            R[i_a, i_s] = ρ(L)
        end
    end

    cs1 = ax.contourf(a_vals, s_vals, transpose(R), alpha=0.5)
    ctr1 = ax.contour(a_vals, s_vals, transpose(R), levels=[1.0])
    plt.clabel(ctr1, inline=1, fontsize=13)
    plt.colorbar(cs1, ax=ax) #, format="%.6f")

    ax.set_xlabel(L"a", fontsize=fontsize)
    ax.set_ylabel(L"s", fontsize=fontsize)

    #ax.set_xticks((minimum(a_vals), maximum(a_vals)))
    #ax.set_yticks((minimum(s_vals), maximum(s_vals)))

    if savefig
        fig.savefig(figname)
    end
    plt.show()
end

File: code/py/discount_spec_rad.py
import numpy as np
from numba import njit
import matplotlib.pyplot as plt
from quantecon import tauchen

fontsize = 18

# Spectral radius
@njit
def ρ(A):
    return np.max(np.abs(np.linalg.eigvals(A)))

def plot_contours(savefig=False, figname="../figures_py/discount_spec_rad.png"):
    fig, ax = plt.subplots(figsize=(8, 4.2))
    grid_size = 50
    n = 6
    a_vals = np.linspace(0.1, 0.95, grid_size)
    s_vals = np.linspace(0.1, 0.32, grid_size)
    μ = 0.96

    R = np.zeros((grid_size, grid_size))
    L = np.zeros((n, n))

    for i_a, a in enumerate(a_vals):
        for i_s, s in enumerate(s_vals):
            mc = tauchen(n, a, np.sqrt(1 - a**2) * s, (1 - a) * μ)
            z_vals, Q = mc.state_values, mc.P
            β_vals = z_vals # np.maximum(z_vals, 0)
            L = β_vals * Q
            R[i_a, i_s] = ρ(L)

    cs1 = ax.contourf(a_vals, s_vals, R.T, alpha=0.5)
    ctr1 = ax.contour(a_vals, s_vals, R.T, levels=[1.0])
    plt.clabel(ctr1, inline=1, fontsize=13)
    plt.colorbar(cs1, ax=ax)

    ax.set_xlabel(r"$a$", fontsize=fontsize)
    ax.set_ylabel(r"$s$", fontsize=fontsize)

    if savefig:
        fig.savefig(figname)
    plt.show()


plot_contours(True)
